js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+    | GetImplicitPkFieldCompound<O>
+    | GetImplicitPkFieldSingle<O>;
+
+interface ViewModelOptions<O extends FieldsMapping> {
+    baseClass?: ViewModelConstructor<FieldsMapping>;
+    fields: O;
+    pkFieldName?: null | undefined | string | string[];
+    getImplicitPkField?: null | undefined | GetImplicitPkField<O>;
+}
+
+interface ViewModelOptionsPkFieldNameSingle<O extends FieldsMapping> extends ViewModelOptions<O> {
+    pkFieldName: string;
+}
+interface ViewModelOptionsPkFieldNameCompound<O extends FieldsMapping> extends ViewModelOptions<O> {
+    pkFieldName: string[];
+}
+
+interface ViewModelOptionsGetImplicitPkFieldSingle<O extends FieldsMapping>
+    extends ViewModelOptions<O> {
+    getImplicitPkField: GetImplicitPkFieldSingle<O>;
+}
+
+interface ViewModelOptionsGetImplicitPkFieldCompound<O extends FieldsMapping>
+    extends ViewModelOptions<O> {
+    getImplicitPkField: GetImplicitPkFieldCompound<O>;
+}
+
+function defineRequiredGetter(base: {}, name: string, errorMessage: string): void {
+    Object.defineProperty(base, name, {
+        configurable: true,
+        get(): string {
+            throw new Error(errorMessage);
+        },
+        // Setter only appears to be required in node when running tests - in all browser engines
+        // I've tried defining the static property on the class overrides the property and does not
+        // call the setter but this probably also depends on the build setup.
+        set(label): void {
+            Object.defineProperty(base, name, {
+                value: label,
+            });
+        },
+    });
+}
+
+/**
+ * Generate a label for a field based on its name
+ */
+function generateFieldLabel(name: string): string {
+    // Inner startCase splits into words and lowercases it:
+    // EMAIL_ADDRESS => email address
+    // Outer one converts first letter of each word:
+    // email address => Email Address
+    return startCase(startCase(name).toLowerCase());
+}
+
+function bindFields<O extends FieldsMapping>(fields: O, bindTo: ViewModelConstructor<O>): O {
+    const newFields = Object.entries(fields).reduce((acc, [fieldName, field]) => {
+        acc[fieldName] = field.clone();
+        acc[fieldName].parent = bindTo;
+        acc[fieldName].name = fieldName;
+        if (acc[fieldName].label === undefined) {
+            acc[fieldName].label = generateFieldLabel(fieldName);
+        }
+        return acc;
+    }, {});
+    return freezeObject(newFields) as O;
+}
+
+function defaultGetImplicitPkField<O extends FieldsMapping>(
+    model: ViewModelConstructor<O>,
+    fields: O
+): ['id', NumberField] {
+    return ['id', fields.id || new NumberField()];
+}
+
+// Overloads here are so we can more accurately type the primary key and fields (specifically for default case we can
+// add the implicit 'id' field that will be created)
+export default function ViewModelFactory<O extends FieldsMapping>(
+    options: ViewModelOptionsPkFieldNameSingle<O>
```

Comment:
hmmm actually now I think about usage a bit more this should probably be split into two arguments - `fields` which is always provided and then the options.

pretty minor change so I'll address that later if you are in agreement

Reply:
actually I'm going to do it now